
INSERT INTO odestatic.variant_type (vt_ref_id, vt_effect, vt_description)
VALUES
    (NULL, 'unknown', 'Unkown variant effect'),
    ('SO:0001537', 'structural_variant', 'A sequence variant that changes one or more sequence features.'),
    ('SO:0002160', 'sequence_length_variant', 'A sequence variant that changes the length of one or more sequence features.'),
    ('SO:0001881', 'feature_translocation', 'A sequence variant, caused by an alteration of the genomic sequence, where the structural change, a translocation, is greater than the extent of the underlying genomic features.'),
    ('SO:0001880', 'feature_amplification', 'A sequence variant, caused by an alteration of the genomic sequence, where the structural change, an amplification of sequence, is greater than the extent of the underlying genomic features.'),
    ('SO:0001878', 'feature_variant', 'A sequence variant that falls entirely or partially within a genomic feature.'),
    ('SO:0001879', 'feature_ablation', 'A sequence variant, caused by an alteration of the genomic sequence, where the deletion, is greater than the extent of the underlying genomic features.'),
    ('SO:0001882', 'feature_fusion', 'A sequence variant, caused by an alteration of the genomic sequence, where a deletion fuses genomic features.'),
    ('SO:0002161', 'short_tandem_repeat_change', 'A sequence variant where the copies of a short tandem repeat (STR) feature are either contracted or expanded.'),
    ('SO:0001563', 'copy_number_change', 'A sequence variant where copies of a feature (CNV) are either increased or decreased.'),
    ('SO:0001883', 'transcript_translocation', 'A feature translocation where the region contains a transcript.'),
    ('SO:0001884', 'regulatory_region_translocation', 'A feature translocation where the region contains a regulatory region.'),
    ('SO:0001889', 'transcript_amplification', 'A feature amplification of a region containing a transcript.'),
    ('SO:0001891', 'regulatory_region_amplification', 'A feature amplification of a region containing a regulatory region.'),
    ('SO:0001906', 'feature_truncation', 'A sequence variant that causes the reduction of a genomic feature, with regard to the reference sequence.'),
    ('SO:0001628', 'intergenic_variant', 'A sequence variant located in the intergenic region, between genes.'),
    ('SO:0001566', 'regulatory_region_variant', 'A sequence variant located within a regulatory region.'),
    ('SO:0001564', 'gene_variant', 'A sequence variant where the structure of the gene is changed.'),
    ('SO:0001907', 'feature_elongation', 'A sequence variant that causes the extension of a genomic feature, with regard to the reference sequence.'),
    ('SO:0001017', 'silent_mutation', 'A sequence variant that does not affect protein function. Silent mutations may occur in genic ( CDS, UTR, intron etc) and intergenic regions. Silent mutations may have affects on processes such as splicing and regulation.'),
    ('SO:0001893', 'transcript_ablation', 'A feature ablation whereby the deleted region includes a transcript feature.'),
    ('SO:0001894', 'regulatory_region_ablation', 'A feature ablation whereby the deleted region includes a regulatory region.'),
    ('SO:0001890', 'transcript_regulatory_region_fusion', 'A feature fusion where the deletion brings together a regulatory region and a transcript region.'),
    ('SO:0001887', 'regulatory_region_fusion', 'A feature fusion where the deletion brings together regulatory regions.'),
    ('SO:0001886', 'transcript_fusion', 'A feature fusion where the deletion brings together transcript regions.'),
    ('SO:0002163', 'short_tandem_repeat_contraction', 'A short tandem repeat variant containing fewer repeat units than the reference sequence.'),
    ('SO:0002162', 'short_tandem_repeat_expansion', 'A short tandem repeat variant containing more repeat units than the reference sequence.'),
    ('SO:0001912', 'copy_number_decrease', 'A sequence variant where copies of a feature are decreased relative to the reference.'),
    ('SO:0001911', 'copy_number_increase', 'A sequence variant where copies of a feature are increased relative to the reference.'),
    ('SO:0001885', 'TFBS_translocation', 'A feature translocation where the region contains a transcription factor binding site.'),
    ('SO:0001892', 'TFBS_amplification', 'A feature amplification of a region containing a transcription factor binding site.'),
    ('SO:0001822', 'inframe_deletion', 'An inframe non synonymous variant that deletes bases from the coding sequence.'),
    ('SO:0001587', 'stop_gained', 'A sequence variant whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened polypeptide.'),
    ('SO:0001910', 'frameshift_truncation', 'A frameshift variant that causes the translational reading frame to be shortened relative to the reference feature.'),
    ('SO:0001631', 'upstream_gene_variant', 'A sequence variant located 5'' of a gene.'),
    ('SO:0001632', 'downstream_gene_variant', 'A sequence variant located 3'' of a gene.'),
    ('SO:0001986', 'upstream_transcript_variant', 'A feature variant, where the alteration occurs upstream of the transcript TSS.'),
    ('SO:0001987', 'downstream_transcript_variant', ''),
    ('SO:0002074', 'intergenic_1kb_variant', 'A variant that falls in an intergenic region that is 1 kb or less between 2 genes.'),
    ('SO:0002017', 'conserved_intergenic_variant', 'A sequence variant located in a conserved intergenic region, between genes.'),
    ('SO:0001782', 'TF_binding_site_variant', 'A sequence variant located within a transcription factor binding site.'),
    ('SO:0002152', 'genic_downstream_transcript_variant', 'A variant that falls downstream of a transcript, but within the genic region of the gene due to alternately transcribed isoforms.'),
    ('SO:0002153', 'genic_upstream_transcript_variant', 'A variant that falls upstream of a transcript, but within the genic region of the gene due to alternately transcribed isoforms.'),
    ('SO:0001565', 'gene_fusion', 'A sequence variant whereby a two genes have become joined.'),
    ('SO:0001576', 'transcript_variant', 'A sequence variant that changes the structure of the transcript.'),
    ('SO:0001598', 'translational_product_structure_variant', 'A sequence variant within the transcript that changes the structure of the translational product.'),
    ('SO:0001908', 'internal_feature_elongation', 'A sequence variant that causes the extension of a genomic feature from within the feature rather than from the terminus of the feature, with regard to the reference sequence.'),
    ('SO:0001578', 'stop_lost', 'A sequence variant where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript.'),
    ('SO:0001895', 'TFBS_ablation', 'A feature ablation whereby the deleted region includes a transcription factor binding site.'),
    ('SO:0001888', 'TFBS_fusion', 'A fusion where the deletion brings together transcription factor binding sites.'),
    ('SO:0002165', 'trinucleotide_repeat_expansion', 'A short tandem repeat expansion with an increase in a sequence of three nucleotide units repeated in tandem compared to a reference sequence.'),
    ('SO:0001825', 'conservative_inframe_deletion', 'An inframe decrease in cds length that deletes one or more entire codons from the coding sequence but does not change any remaining codons.'),
    ('SO:0001826', 'disruptive_inframe_deletion', 'An inframe decrease in cds length that deletes bases from the coding sequence starting within an existing codon.'),
    ('SO:0001636', '2KB_upstream_variant', 'A sequence variant located within 2KB 5'' of a gene.'),
    ('SO:0001635', '5KB_upstream_variant', 'A sequence variant located within 5KB 5'' of a gene.'),
    ('SO:0001634', '500B_downstream_variant', 'A sequence variant located within a half KB of the end of a gene.'),
    ('SO:0002083', '2KB_downstream_variant', 'A sequence variant located within 2KB 3'' of a gene.'),
    ('SO:0001633', '5KB_downstream_variant', 'A sequence variant located within 5 KB of the end of a gene.'),
    ('SO:0002085', 'unidirectional_gene_fusion', 'A sequence variant whereby two genes, on the same strand have become joined.'),
    ('SO:0002086', 'bidirectional_gene_fusion', 'A sequence variant whereby two genes, on alternate strands have become joined.'),
    ('SO:0001619', 'non_coding_transcript_variant', 'A transcript variant of a non coding RNA gene.'),
    ('SO:0002011', 'intragenic_variant', 'A variant that occurs within a gene but falls outside of all transcript features. This occurs when alternate transcripts of a gene do not share overlapping  sequence.'),
    ('SO:0001968', 'coding_transcript_variant', 'A transcript variant of a protein coding gene.'),
    ('SO:0001627', 'intron_variant', 'A transcript variant occurring within an intron.'),
    ('SO:0001621', 'NMD_transcript_variant', 'A variant in a transcript that is the target of NMD.'),
    ('SO:0001596', 'transcript_secondary_structure_variant', 'A sequence variant within a transcript that changes the secondary structure of the RNA product.'),
    ('SO:0001568', 'splicing_variant', 'A sequence variant that changes the process of splicing.'),
    ('SO:0002075', 'incomplete_transcript_variant', 'A sequence variant that intersects an incompletely annotated transcript.'),
    ('SO:0001791', 'exon_variant', 'A sequence variant that changes exon sequence.'),
    ('SO:0001577', 'complex_transcript_variant', 'A transcript variant with a complex INDEL- Insertion or deletion that spans an exon/intron border or a coding sequence/UTR border.'),
    ('SO:0001603', 'polypeptide_sequence_variant', 'A sequence variant with in the CDS that causes a change in the resulting polypeptide sequence.'),
    ('SO:0001909', 'frameshift_elongation', 'A frameshift variant that causes the translational reading frame to be extended relative to the reference feature.'),
    ('SO:0001821', 'inframe_insertion', 'An inframe non synonymous variant that inserts bases into in the coding sequence.'),
    ('SO:0002088', 'non_coding_transcript_splice_region_variant', 'A transcript variant occurring within the splice region (1-3 bases of the exon or 3-8 bases of the intron) of a non coding transcript.'),
    ('SO:0001620', 'mature_miRNA_variant', 'A transcript variant located with the sequence of the mature miRNA.'),
    ('SO:0001970', 'non_coding_transcript_intron_variant', 'A transcript variant occurring within an intron of a non coding transcript.'),
    ('SO:0002018', 'conserved_intron_variant', 'A transcript variant occurring within a conserved region of an intron.'),
    ('SO:0001969', 'coding_transcript_intron_variant', 'A transcript variant occurring within an intron of a coding transcript.'),
    ('SO:0001597', 'compensatory_transcript_secondary_structure_variant', 'A secondary structure variant that compensate for the change made by a previous variant.'),
    ('SO:0002169', 'splice_polypyrimidine_tract_variant', 'A sequence variant that falls in the polypyrimidine tract at 3'' end of intron between 17 and 3 bases from the end (acceptor -3 to acceptor -17).'),
    ('SO:0001995', 'extended_intronic_splice_region_variant', 'A sequence variant occurring in the intron, within 10 bases of exon.'),
    ('SO:0001630', 'splice_region_variant', 'A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron.'),
    ('SO:0001569', 'cryptic_splice_site_variant', 'A sequence variant causing a new (functional) splice site.'),
    ('SO:0001629', 'splice_site_variant', 'A sequence variant that changes the first two or last two bases of an intron, or the 5th base from the start of the intron in the orientation of the transcript.'),
    ('SO:0001573', 'intron_gain_variant', 'A sequence variant whereby an intron is gained by the processed transcript; usually a result of an alteration of the donor or acceptor.'),
    ('SO:0001572', 'exon_loss_variant', 'A sequence variant whereby an exon is lost from the transcript.'),
    ('SO:0002076', 'incomplete_transcript_3UTR_variant', 'A sequence variant that intersects the 3'' UTR of an incompletely annotated transcript.'),
    ('SO:0002079', 'incomplete_transcript_splice_region_variant', 'A sequence variant that intersects the splice region of an incompletely annotated transcript.'),
    ('SO:0002078', 'incomplete_transcript_intronic_variant', 'A sequence variant that intersects the intron of an incompletely annotated transcript.'),
    ('SO:0002077', 'incomplete_transcript_5UTR_variant', 'A sequence variant that intersects the 5'' UTR of an incompletely annotated transcript.'),
    ('SO:0002080', 'incomplete_transcript_exonic_variant', 'A sequence variant that intersects the exon of an incompletely annotated transcript.'),
    ('SO:0002081', 'incomplete_transcript_CDS', 'A sequence variant that intersects the coding regions of an incompletely annotated transcript.'),
    ('SO:0001792', 'non_coding_transcript_exon_variant', 'A sequence variant that changes non-coding exon sequence in a non-coding transcript.'),
    ('SO:0001580', 'coding_sequence_variant', 'A sequence variant that changes the coding sequence.'),
    ('SO:0001622', 'UTR_variant', 'A transcript variant that is located within the UTR.'),
    ('SO:0001609', 'elongated_polypeptide', 'A sequence variant with in the CDS that causes elongation of the resulting polypeptide sequence.'),
    ('SO:0001606', 'amino_acid_substitution', 'A sequence variant of a codon resulting in the substitution of one amino acid for another in the resulting polypeptide.'),
    ('SO:0001605', 'amino_acid_insertion', 'A sequence variant within a CDS resulting in the gain of an amino acid to the resulting polypeptide.'),
    ('SO:0001604', 'amino_acid_deletion', 'A sequence variant within a CDS resulting in the loss of an amino acid from the resulting polypeptide.'),
    ('SO:0001616', 'polypeptide_fusion', 'A sequence variant that causes a fusion of two polypeptide sequences.'),
    ('SO:0001617', 'polypeptide_truncation', 'A sequence variant of the CD that causes a truncation of the resulting polypeptide.'),
    ('SO:0001823', 'conservative_inframe_insertion', 'An inframe increase in cds length that inserts one or more codons into the coding sequence between existing codons.'),
    ('SO:0001824', 'disruptive_inframe_insertion', 'An inframe increase in cds length that inserts one or more codons into the coding sequence within an existing codon.'),
    ('SO:0002090', '3_prime_UTR_intron_variant', 'A UTR variant of intronic sequence of the 3'' UTR.'),
    ('SO:0002091', '5_prime_UTR_intron_variant', 'A UTR variant of intronic sequence of the 5'' UTR.'),
    ('SO:0002084', 'exonic_splice_region_variant', 'A sequence variant in which a change has occurred within the exonic region of the splice site, 1-2 bases from boundary.'),
    ('SO:0002170', 'splice_donor_region_variant', 'A sequence variant that falls in the region between the 3rd and 6th base after splice junction (5'' end of intron).'),
    ('SO:0002088', 'non_coding_transcript_splice_region_variant', 'A transcript variant occurring within the splice region (1-3 bases of the exon or 3-8 bases of the intron) of a non coding transcript.'),
    ('SO:0001571', 'cryptic_splice_donor', 'A sequence variant whereby a new splice site is created due to the activation of a new donor.'),
    ('SO:0001570', 'cryptic_splice_acceptor', 'A sequence variant whereby a new splice site is created due to the activation of a new acceptor.'),
    ('SO:0001575', 'splice_donor_variant', 'A splice variant that changes the 2 base pair region at the 5'' end of an intron.'),
    ('SO:0001574', 'splice_acceptor_variant', 'A splice variant that changes the 2 base region at the 3'' end of an intron.'),
    ('SO:0001787', 'splice_donor_5th_base_variant', 'A sequence variant that causes a change at the 5th base pair after the start of the intron in the orientation of the transcript.'),
    ('SO:0002082', 'incomplete_transcript_coding_splice_variant', 'A sequence variant that intersects the coding sequence near a splice region of an incompletely annotated transcript.'),
    ('SO:0001818', 'protein_altering_variant', 'A sequence_variant which is predicted to change the protein encoded in the coding sequence.'),
    ('SO:0001819', 'synonymous_variant', 'A sequence variant where there is no resulting change to the encoded amino acid.'),
    ('SO:0001590', 'terminator_codon_variant', 'A sequence variant whereby at least one of the bases in the terminator codon is changed.'),
    ('SO:0001582', 'initiator_codon_variant', 'A codon variant that changes at least one base of the first codon of a transcript.'),
    ('SO:0001624', '3_prime_UTR_variant', 'A UTR variant of the 3'' UTR.'),
    ('SO:0001623', '5_prime_UTR_variant', 'A UTR variant of the 5'' UTR.'),
    ('SO:0001610', 'elongated_polypeptide_C_terminal', 'A sequence variant with in the CDS that causes elongation of the resulting polypeptide sequence at the C terminus.'),
    ('SO:0001611', 'elongated_polypeptide_N_terminal', 'A sequence variant with in the CDS that causes elongation of the resulting polypeptide sequence at the N terminus.'),
    ('SO:0001607', 'conservative_amino_acid_substitution', 'A sequence variant of a codon causing the substitution of a similar amino acid for another in the resulting polypeptide.'),
    ('SO:0001608', 'non_conservative_amino_acid_substitution', 'A sequence variant of a codon causing the substitution of a non conservative amino acid for another in the resulting polypeptide.'),
    ('SO:0001650', 'inframe_variant', 'A sequence variant which does not cause a disruption of the translational reading frame.'),
    ('SO:0001589', 'frameshift_variant', 'A sequence variant which causes a disruption of the translational reading frame, because the number of nucleotides inserted or deleted is not a multiple of three.'),
    ('SO:0001567', 'stop_retained_variant', 'A sequence variant where at least one base in the terminator codon is changed, but the terminator remains.'),
    ('SO:0001578', 'stop_lost', 'A sequence variant where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript.'),
    ('SO:0002012', 'start_lost', 'A codon variant that changes at least one base of the canonical start codon.'),
    ('SO:0002019', 'start_retained_variant', 'A sequence variant where at least one base in the start codon is changed, but the start remains.'),
    ('SO:0002089', '3_prime_UTR_exon_variant', 'A UTR variant of exonic sequence of the 3'' UTR.'),
    ('SO:0002090', '3_prime_UTR_intron_variant', 'A UTR variant of intronic sequence of the 3'' UTR.'),
    ('SO:0002016', '3_prime_UTR_elongation', 'A sequence variant that causes the extension of 3'' UTR, with regard to the reference sequence.'),
    ('SO:0002015', '3_prime_UTR_truncation', 'A sequence variant that causes the reduction of a the 3'' UTR  with regard to the reference sequence.'),
    ('SO:0002092', '5_prime_UTR_exon_variant', 'A UTR variant of exonic sequence of the 5'' UTR.'),
    ('SO:0002091', '5_prime_UTR_intron_variant', 'A UTR variant of intronic sequence of the 5'' UTR.'),
    ('SO:0001983', '5_prime_UTR_premature_start_codon_variant', 'A 5'' UTR variant where a premature start codon is introduced, moved or lost.'),
    ('SO:0002013', '5_prime_UTR_truncation', 'A sequence variant that causes the reduction of a the 5''UTR  with regard to the reference sequence.'),
    ('SO:0002014', '5_prime_UTR_elongation', 'A sequence variant that causes the extension of 5'' UTR, with regard to the reference sequence.'),
    ('SO:0001612', 'elongated_in_frame_polypeptide_C_terminal', 'A sequence variant with in the CDS that causes in frame elongation of the resulting polypeptide sequence at the C terminus.'),
    ('SO:0001613', 'elongated_out_of_frame_polypeptide_C_terminal', 'A sequence variant with in the CDS that causes out of frame elongation of the resulting polypeptide sequence at the C terminus.'),
    ('SO:0001614', 'elongated_in_frame_polypeptide_N_terminal_elongation', 'A sequence variant with in the CDS that causes in frame elongation of the resulting polypeptide sequence at the N terminus.'),
    ('SO:0001615', 'elongated_out_of_frame_polypeptide_N_terminal', 'A sequence variant with in the CDS that causes out of frame elongation of the resulting polypeptide sequence at the N terminus.'),
    ('SO:0001626', 'incomplete_terminal_codon_variant', 'A sequence variant where at least one base of the final codon of an incompletely annotated transcript is changed.'),
    ('SO:0001820', 'inframe_indel', 'A coding sequence variant where the change does not alter the frame of the transcript.'),
    ('SO:0001992', 'nonsynonymous_variant', 'A non-synonymous variant is an inframe, protein altering variant, resulting in a codon change.'),
    ('SO:0001909', 'frameshift_elongation', 'A frameshift variant that causes the translational reading frame to be extended relative to the reference feature.'),
    ('SO:0001910', 'frameshift_truncation', 'A frameshift variant that causes the translational reading frame to be shortened relative to the reference feature.'),
    ('SO:0001593', 'minus_2_frameshift_variant', ''),
    ('SO:0001592', 'minus_1_frameshift_variant', 'A sequence variant which causes a disruption of the translational reading frame, by shifting one base ahead.'),
    ('SO:0001591', 'frame_restoring_variant', 'A sequence variant that reverts the sequence of a previous frameshift mutation back to the initial frame.'),
    ('SO:0001595', 'plus_2_frameshift_variant', ''),
    ('SO:0001594', 'plus_1_frameshift_variant', 'A sequence variant which causes a disruption of the translational reading frame, by shifting one base backward.'),
    ('SO:0001988', '5_prime_UTR_premature_start_codon_gain_variant', 'A 5'' UTR variant where a premature start codon is gained.'),
    ('SO:0001989', '5_prime_UTR_premature_start_codon_loss_variant', 'A 5'' UTR variant where a premature start codon is lost.'),
    ('SO:0001990', 'five_prime_UTR_premature_start_codon_location_variant', 'A 5'' UTR variant where a premature start codon is moved.'),
    ('SO:0001822', 'inframe_deletion', 'An inframe non synonymous variant that deletes bases from the coding sequence.'),
    ('SO:0001821', 'inframe_insertion', 'An inframe non synonymous variant that inserts bases into in the coding sequence.'),
    ('SO:0002012', 'start_lost', 'A codon variant that changes at least one base of the canonical start codon.'),
    ('SO:0001587', 'stop_gained', 'A sequence variant whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened polypeptide.'),
    ('SO:0001578', 'stop_lost', 'A sequence variant where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript.'),
    ('SO:0001583', 'missense_variant', 'A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved.'),
    ('SO:0001825', 'conservative_inframe_deletion', 'An inframe decrease in cds length that deletes one or more entire codons from the coding sequence but does not change any remaining codons.'),
    ('SO:0001826', 'disruptive_inframe_deletion', 'An inframe decrease in cds length that deletes bases from the coding sequence starting within an existing codon.'),
    ('SO:0001823', 'conservative_inframe_insertion', 'An inframe increase in cds length that inserts one or more codons into the coding sequence between existing codons.'),
    ('SO:0001824', 'disruptive_inframe_insertion', 'An inframe increase in cds length that inserts one or more codons into the coding sequence within an existing codon.'),
    ('SO:0001585', 'conservative_missense_variant', 'A sequence variant whereby at least one base of a codon is changed resulting in a codon that encodes for a different but similar amino acid. These variants may or may not be deleterious.'),
    ('SO:0001586', 'non_conservative_missense_variant', 'A sequence variant whereby at least one base of a codon is changed resulting in a codon that encodes for an amino acid with different biochemical properties.'),
    ('SO:0002008', 'rare_amino_acid_variant', 'A sequence variant whereby at least one base of a codon encoding a rare amino acid is changed, resulting in a different encoded amino acid.'),
    ('SO:0002010', 'pyrrolysine_loss', 'A sequence variant whereby at least one base of a codon encoding pyrrolysine is changed, resulting in a different encoded amino acid.'),
    ('SO:0002009', 'selenocysteine_loss', 'A sequence variant whereby at least one base of a codon encoding selenocysteine  is changed, resulting in a different encoded amino acid.');

